### ISSUE TITLE
[RFC] Lowering `aten.index_add_` through the TM Tensor Dialect

### ISSUE THREAD

**Issue by lorenzbc** [llvm/torch-mlir#708]
Hey @ramiro050 and @yihonglyu,

Trying to move `aten.index_add_` so that it goes through the TM Tensor Dialect 

Inspired by [this PR lowering aten.bincount](https://github.com/llvm/torch-mlir/pull/574/commits/f818ce22f3999e28b05455b188aaa24dc9ee1798), here's my current plan: 

1. Add `AtenIndexAdd` to `e2e_testing/torchscript/basic.py`
2. Define `Toch_AtenIndexAddOp` in `include/torch-mlir/Dialect/Torch/IR/GeneratedAtenOps.td`
3. Create a `AtenIndexAddOp` class in `lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp`
-- I'm also going to the op under `runOnOperation` within the `ConvertTorchToTmTensor` class
4. Add the op to the conditional statement in `lib/Dialect/Torch/Transforms/RefineTypes.cpp` 
5. Emit the op under `python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py`

Would love any feedback! Am I thinking about the sequence correctly? 

Thank you in advance, 

Lorenzo

**Comment by ramiro050** [comment]
Looks good. I would perform step 5 before anything else, so that you're able to access the op in C++. Also, is there a reason why we cannot decompose this op into `index_put` with the `accumulate` flag set to `True`?

**Comment by lorenzbc** [comment]
Looping the convo @ramiro050 and I had had offline

`index_add_` could potentially become an instance of `index_put`

[The documentation](https://pytorch.org/docs/stable/generated/torch.Tensor.index_put_.html) statutes that the `accumulate` flag turns the op from `tensor[index] = value` to `tensor[index] += value` 

This behavior makes sense but imo is not intuitive (as evidenced by the fact that I didn't know `index_put` could add before Ramiro pointed it out), so creating a separate op but writing as an instance of `index_put` might still be worth it

**Comment by ramiro050** [comment]
Yeah, I think a good first step would be to have some pseudo-code for what the decomposition would look like because after looking at the `index_add` op more closely, it is not trivial to decompose to `index_put`. Some careful broadcasting has to be performed. That would give us a good understanding of if there are limitations to the decomposition approach or not.

**Comment by lorenzbc** [comment]
Hey Ramiro, 

Looking at [the index_put.py test](https://github.com/llvm/torch-mlir/blob/6b637a9fd91022f18eb272fac5426bc93034f901/python/torch_mlir_e2e_test/test_suite/index_put.py) on the torch-mlir side, it seems that `IndexPutOneDimFloatAccumulateModule` and `IndexPutOneDimIntAccumulateModule` both rely on `torch.ops.aten.index_put` with the accumulated flag set to True

In [TorchToTMTensor.cpp](https://github.com/lorenzbc/torch-mlir/blob/main/lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp) there's a `ConvertValsemVariantAtenIndexPutImplOp` function that seems to be converting the op to the TM Dialect

Can't I do a similar thing and define `index_add` through index_put.py and then convert it through a function in TorchToTMTensor.cpp?

**Comment by cathyzhyi** [comment]
Hey @lorenzbc, the code for `index_put.py` is for testing so `torch.ops.aten.index_add` must there.  Potentially, you might be able to decompose `torch.ops.aten.index_add` into `torch.ops.aten.index_put` in `lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp` but as Ramiro mentioned in https://github.com/llvm/torch-mlir/issues/708#issuecomment-1081289430 the first step is to double check the pseudo code to have better understanding if it's doable. If it's doable, then `torch.ops.aten.index_add` can be converted to `torch.ops.aten.index_put` and nothing need to be added to https://github.com/lorenzbc/torch-mlir/blob/main/lib/Conversion/TorchToTMTensor/TorchToTMTensor.cpp. If it's not doable, the lowering from `torch.ops.aten.index_add` to tmtensor/linalg ops need to be implemented in TorchToTMTensor.cpp.

**Comment by lorenzbc** [comment]
hey @cathyzhyi

I talked to @ramiro050 last Friday to figure out how to reshape the updates tensor so that it won't throw an error during broadcasting. 

This is the reference example in pytorch 
![image](https://user-images.githubusercontent.com/101747922/162069765-b80fa6b2-9d7a-4098-b680-d58480941782.png)

Here are the 3 conditions for the reshaping: 

1. `updates` needs to have the same rank as `input`
2. `updates.shape(dim) == len(indices)` 
3. Every new dimension added to `updates` is of size 1 

I'm about to write a testcase to see which ops are used in the IR. My idea right now is to use those ops within `DecomposeComplexOps.cpp` as in [line 1539](https://github.com/llvm/torch-mlir/blob/main/lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp#L1539)
